src/lib/device-online-state/index.ts

```diff
@@ -0,0 +1,280 @@
+				}
+			})
+			.then((msg: any) => {
+				const { id, message } = msg;
+
+				return Bluebird.try(() => {
+					const { uuid, nextState } = JSON.parse(message);
+
+					// raise and event for the state change...
+					switch (nextState) {
+						case DeviceOnlineStates.Timeout:
+							this.scheduleChangeOfStateForDevice(
+								uuid,
+								DeviceOnlineStates.Timeout,
+								DeviceOnlineStates.Offline,
+								15, // put the device into a timeout state for 15 sec if it misses it's scheduled heartbeat window... then mark as offline
+							);
+							this.deviceTimeout(uuid);
+							break;
+						case DeviceOnlineStates.Offline:
+							this.deviceOffline(uuid);
+							break;
+						default:
+							// nothing should be raised here...
+							break;
+					}
+				}).then(() =>
+					this.rsmq.deleteMessage({
+						qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+						id: id,
+					}),
+				);
+			})
+			.catch(NoMessagesToConsumeSignal, noop)
+			.catch((err: Error) => {
+				captureException(
+					err,
+					'DeviceStateManager ==> Error when consuming queue',
+				);
+			})
+			.then(() => {
+				// schedule the next consume...
+				setTimeout(() => {
+					this.consume();
+				}, delayBeforeNextConsume);
+			});
+	}
+
+	private scheduleChangeOfStateForDevice(
+		uuid: string,
+		currentState: DeviceOnlineStates,
+		nextState: DeviceOnlineStates,
+		inSeconds: number,
+	) {
+		// remove the old queued state...
+		return this.redis
+			.getAsync(`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`)
+			.then(value => {
+				const { id } = JSON.parse(value) as { id: string };
+
+				if (id) {
+					return this.rsmq
+						.deleteMessage({
+							qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+							id,
+						})
+						.return();
+				}
+			})
+			.catch(() => noop) // ignore errors when deleting the old queued state and queue the new one...
+			.then(() => {
+				return this.rsmq.sendMessage({
+					qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+					message: JSON.stringify({
+						uuid,
+						nextState,
+					}),
+					delay: inSeconds,
+				});
+			})
+			.then(id => {
+				return this.redis
+					.setAsync(
+						`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`,
+						JSON.stringify({
+							id,
+							currentState,
+						}),
+					)
+					.then(() =>
+						this.redis.expireAsync(
+							`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`,
+							inSeconds + 5, // add a small 5 second buffer to allow some wiggle room on top of the timeout window...
+						),
+					);
+			});
+	}
+
+	public captureEventFor(uuid: string, timeoutSeconds: number) {
+		// see if we already have a queued state for this device...
+		return this.redis
+			.getAsync(`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`)
+			.then(value => {
+				const { id, currentState } = JSON.parse(value);
+
+				if (!id || DeviceOnlineStates.Online !== currentState) {
+					this.deviceOnline(uuid);
+				}
+			})
+			.catch(() => noop)
```

Comment:
Do you really mean to return a noop function here?

Reply:
Again, this is getting an ID for a message which is queued for later. If it cannot find one, then it could be an error, but it is more than likely that this device has just come online from being offline for a while. We couldn't determine if this is a real error or not, so just drop it and move on.